### ISSUE TITLE
pypi upload of workflows

### ISSUE THREAD

**Issue by Hoohm** [snakemake-workflows/docs#12]
Hello fellow snakemakers!

I was thinking about how I could get the installation and use of our workflow to be even easier and thought of using pypi as another repo. This would mean that the workflow would now also be a python package.

It comes with a few neat things, one of them is that you can run a workflow from anywhere as an executable passing all arguments to snakemake as usual.

Instead of having to be in the workflow folder and run `snakemake --directory WORKINGDIR` you can run `workflow_name --directory WORKINGDIR` from anywhere and it will run the same thing.

Here is the code of the __main__.py file

```
import sys
import subprocess
import os

def main():
	arguments = ['snakemake']
	arguments.extend(sys.argv[1:])
	subprocess.run(args=arguments, 	cwd=os.path.join(os.path.dirname(__file__),'..'))


if __name__=='__main__':
	main()
```

Another thing that I wanted to implement is a "prepare" functionality which would be an interactive question/answer in the prompt to generate `samples.csv` and configuration files. In our workflow, I would, for example, ask "which chemistry have you used for your experiment" giving a list of available choices directly in the prompt. This would then write the correct chemistry directly into the config.yaml instead of relying on people to write it themselves. This would ensure that there are no spelling mistakes and check for int types for example.

This preparing was always "odd" in a workflow because normally you rely on having everything setup before running it. Although the new checkpoints might fit the bill?

One potential issue with the executable approach is that it might become confusing to the users if you have both the executable and the classic workflow available and the fact that the workflow would be in the lib dir of python and not cloned somewhere.

Since this is a big/strange shift I wanted to have your opinion before moving forward with this experiment. Let me know if I'm missing any crucial problems.

Best wishes

**Comment by johanneskoester** [comment]
Hi,
thanks for sharing this idea. I have thought about that a bit in the past. Here are some thoughts:

* I think it is very important to "enforce" people having the workflow code and configuration on disk together with the data. Otherwise, you loose reproducibility and it is harder to keep track of what was actually executed. This does not contradict your concept in general though.
* Validation of config files is already possible now. It works via JSON schemas and the validate function. See [here](https://github.com/snakemake-workflows/dna-seq-gatk-variant-calling/blob/master/rules/common.smk#L8) for an example.

So what remains is the question whether it would be nice to have a command line tool that allows to grab a workflow from github, use the schema to ask users questions that fill the config file and then stores everything on disk for execution with Snakemake. This is indeed a very nice idea. I think such a tool should not be a pypi package per workflow, but rather a general tool that works for any workflow that is hosted in a git repo and follows the best practice layout. Such a tool could also be written in a way that the user can choose whether he wants to answer the questions at the CLI or in a webpage.

What do you think?

**Comment by Hoohm** [comment]
Hi and thanks for your input.


Unique or multiple clones
--------------------------------------
Concerning your first point, I would have thought that:

* The version of the workflow, as well as all fixed package versions
* Hardcoded and unique to experiment config files

are the crucial aspects to make it fully reproducible.

In fact, I'm only using one "copy" of the workflow and use --directory to run on different experiments. This is just to reduce overhead when I update something in the workflow. I would have to change it in each experiment folder which is a big overhead or run a whole round of push/pull for each one.

I'm curious of your thought on this because I imagined giving out the config files and the version of the workflow when publishing some experiments. Any other user would have to download the specific version and run the whole thing with the given config files. But I don't imagine uploading the scripts, rules and/or the .snakemake folder.


General purpose workflow manager
---------------------------
The JSON schemas is definitely a good entry point for data validation and could be used in an interactive way.

A general tool could definitely fit the bill.

Something like:

`workflow-manager NAME_OF_WORKFLOW init TARGET_FOLDER` would clone the workflow locally and run the "init" function, checking requirements with the JSON schema or maybe a custom function that could be stored in the "scripts" folder of a workflow.

Once this is done:

`workflow-manager NAME_OF_WORKFLOW run TARGET_FOLDER`

PS: I can see the CLI interaction but I have no clue on how to make it via a webinterface

**Comment by johanneskoester** [comment]
# Point 1
Yes, in certain scenarios having one codebase that is exeucted in many workdirs can work. However, you rely on the fact that the code never has to be changed. Otherwise you end up with workdirs that do not match the central codebase anymore. Everything just becomes harder to track. Also it can happen that you need to improve or adapt things for one of the "sub"-projects. All of these are problems that are already solved by git, and using it to maintain the code inside of the workdir. When you change something, you branch. When you want to propagate it to other project instances, you push and pull, and so on. No need to come up with a new solution. Also, for publishing a workflow, publishing the code is crucial, because you never know whether some central codebase will be accessible for all times.

# Point 2
I would vote for

```
snakemake-deploy <NAME_OR_GIT_URL>
```
which would then ask questions for filling the config file. That's all that is needed. I am not a fan of hiding Snakemake behind some kind of run wrapper. You would just end up replicating the Snakemake command line interface. If you feel like the options of the Snakemake CLI are too complicated, let us rather try to fix this (e.g., there could be a "simple", and and "expert" CLI help). For unexperienced users, snakemake-deploy could print out what are currently the contents of the readmes in the repos we have here.

**Comment by Hoohm** [comment]
Point1
-------------------------------------
You make very good points here, keeping the code and the data together is definitely better in terms of reproducibility. Let's keep it this way then.

Point2
------------------------------------
I think you're right about the obfuscation issue. It's something that also bothered me but I couldn't figure out a good way to deal with this.

I'm gonna try and test out a few ideas in a package in the following weeks.

Regarding the snakemake CLI complexity, the help list is long and I'd definitely enjoy a "simple" but I'm not sure how it's possible to make it simpler to find the command the users might be looking for.
